### PR TITLE
feat(pid_longitudinal_controller): use Integrated error when vehicle is stopped

### PR DESCRIPTION
**Pull request by danielsanchezaran**
## Description

Related to this [Issue](https://github.com/autowarefoundation/autoware.universe/issues/5391)

The goal is to make the PID use the integrative gain to offset possible slope compensation overshoots. The reason this is not currently done is shown on the [documentation](https://github.com/autowarefoundation/autoware.universe/blob/main/control/pid_longitudinal_controller/README.md#pid-control): 

"Also, the integral term is not accumulated when the vehicle is stopped. This is to prevent unintended accumulation of the integral term in cases such as Autoware assumes that the vehicle is engaged, but an external system has locked the vehicle to start."

But I believe that, by tuning the [I effort parameters](https://github.com/autowarefoundation/autoware_launch/blob/c5afec100128e0456aa1b8589187b97b56e0ebe4/autoware_launch/config/control/trajectory_follower/longitudinal_controller.param.yaml#L30):

    max_i_effort: 0.3
    min_i_effort: -0.3
The maximum acceleration caused by the I controller can be tuned to be confortable & safe. So, the velocity limit is not necessary.

## Tests performed

evaluator tests: [TIER4 internal link](https://evaluation.ci.tier4.jp/evaluation/reports/9ae64f05-bc24-5da8-93ab-56531f56d494?project_id=prd_jt) No degradation when compared to today’s baseline [TIER4 internal link](https://evaluation.tier4.jp/evaluation/reports/5f1c1607-9eb6-55e4-942d-87f757294012?project_id=prd_jt)
## Notes for reviewers
Requires launch changes: https://github.com/autowarefoundation/autoware_launch/pull/698

## Effects on system behavior

The use of the integrative gain is no longer limited by the vehicle speed for the PID controller, and the ego can now start on downward slopes.

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [x] I've confirmed the [contribution guidelines].
- [x] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [ ] The PR follows the [pull request guidelines].
- [ ] The PR has been properly tested.
- [ ] The PR has been reviewed by the code owners.

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [ ] There are no open discussions or they are tracked via tickets.
- [ ] The PR is ready for merge.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

### CODE REVIEW

**Comment by takayuki5168** [comment]
LGTM. Horibe-san will review as well.

**Comment by TakaHoribe** [comment]
@danielsanchezaran Thank you for the PR!

I agree with enabling integration during stops. However, there are several considerations we need to address.

Even in AUTONOMOUS mode, it takes time for a vehicle to start moving. Therefore, the integral term will inevitably accumulate unintentionally during the start. This could have a bad effect on the system, making parameter tuning difficult. When stopped, I think we need a method such as enabling integration only if the vehicle doesn't start moving after a certain period (e.g. 2 seconds).

Additionally, vehicle speed sensors often perform poorly at low speeds, frequently returning zero speed. To avoid unintended integral processing in such vehicles, it remains necessary to have the option to disable integration at low speeds.

**Comment by danielsanchezaran** [comment]
Not sure why the issue was closed, I did not do it.

**Comment by danielsanchezaran** [comment]
With these new changes, the error integration does not kick off until a threshold 
https://github.com/autowarefoundation/autoware.universe/assets/25967964/08fad25b-af3f-4037-a175-3bd4765431c0
There is a parameter to set the threshold limit and a parameter to enable or disable the feature altogether @TakaHoribe @takayuki5168  

Note: the changes to autoware_launch are necessary: https://github.com/autowarefoundation/autoware_launch/pull/698

**Comment by danielsanchezaran** [comment]
> @danielsanchezaran Thank you for the PR!
> 
> I agree with enabling integration during stops. However, there are several considerations we need to address.
> 
> Even in AUTONOMOUS mode, it takes time for a vehicle to start moving. Therefore, the integral term will inevitably accumulate unintentionally during the start. This could have a bad effect on the system, making parameter tuning difficult. When stopped, I think we need a method such as enabling integration only if the vehicle doesn't start moving after a certain period (e.g. 2 seconds).
> 
> Additionally, vehicle speed sensors often perform poorly at low speeds, frequently returning zero speed. To avoid unintended integral processing in such vehicles, it remains necessary to have the option to disable integration at low speeds.

I believe the new changes satisfy these requirements. Could you please take a look again?

**Comment by takayuki5168** [comment]
I leave this PR's review to @TakaHoribe-san.

File: control/pid_longitudinal_controller/src/pid_longitudinal_controller.cpp [hunk 1]
```diff
@@ -959,8 +970,15 @@ double PidLongitudinalController::applyVelocityFeedback(
...
-    (std::abs(current_vel) > m_current_vel_threshold_pid_integrate) && is_under_control;
+
+  const bool vehicle_is_moving = std::abs(current_vel) > m_current_vel_threshold_pid_integrate;
+  const double time_under_control = getTimeUnderControl();
+  const bool vehicle_is_stuck =
+    m_enable_integration_at_low_speed
+      ? !vehicle_is_moving && time_under_control > m_time_threshold_before_pid_integrate
+      : false;
+
+  const bool enable_integration = (vehicle_is_moving || vehicle_is_stuck) && is_under_control;
```
Comment (TakaHoribe) [inline control/pid_longitudinal_controller/src/pid_longitudinal_controller.cpp:981]: The status of whether a vehicle is stuck should not depend on the flag: `m_enable_integration_at_low_speed`?

```suggestion
  const bool vehicle_is_stuck =
     !vehicle_is_moving && time_under_control > m_time_threshold_before_pid_integrate;

  const bool enable_integration = (vehicle_is_moving || (m_enable_integration_at_low_speed && vehicle_is_stuck)) && is_under_control;
```

**Review by TakaHoribe — COMMENTED** [review]
@danielsanchezaran LGTM. Two minor comments:

- Would you write a background of this feature on the README? Maybe new users don't know why the time_threshold is necessary. 
- One suggestion below

File: control/pid_longitudinal_controller/README.md [hunk 2]
```diff
@@ -212,6 +217,8 @@ AutonomouStuff Lexus RX 450h for under 40 km/h driving.
 | time_threshold_before_pid_integration | double | How much time without the vehicle moving must past to enable PID error integration. [s]                                                                            | 5.0           |
 | brake_keeping_acc                     | double | If `enable_brake_keeping_before_stop` is true, a certain acceleration is kept during DRIVE state before the ego stops [m/s^2] See [Brake keeping](#brake-keeping). | 0.2           |
 
+Note:
+
```
Comment (TakaHoribe) [inline control/pid_longitudinal_controller/README.md:221]: ?? Should it be removed?

**Review by TakaHoribe — APPROVED** [review]
LGTM! (but one odd part in README?)

**Pull request merged**